### ISSUE TITLE
Plugin doesn't work. Panasonic changed API. UPDATE - will work soon!!

### ISSUE THREAD

**Issue by mkz212** [homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform#330]
# Plugin works again!
Panasonic has changed the API - again. Thanks to @nanos and @heldchen and many other users, plugin it is working again! Beta 23 is out and soon will be public release!

**Comment by hubert3** [comment]
Really wish there was a local interface like there is on my Mitsubishi AC, Cloud only sucks

Can you use help with anything (http proxy logs of running the original app?)

Have they added some new kind of auth or changed headers again?

**Comment by mkz212** [comment]
> Really wish there was a local interface like there is on my Mitsubishi AC, Cloud only sucks
> 
> Can you use help with anything (http proxy logs of running the original app?)
> 
> Have they added some new kind of auth or changed headers again?

Unfortunately, I have no idea how to read the Panasonic API at all. I know that others use the Panasonic Comfort Cloud app for Android (and I don't even have Android).

There is a lot of discussion here: https://github.com/sockless-coding/panasonic_cc/issues/191

From what one person wrote, not only the basic url has changed. Perhaps they also require 2FA (but for this we are ready, to generate a 6-digit code).

**Comment by mkz212** [comment]
There is already a working API but in Python. We need to translate it into Javascript. Does anyone know about it?

https://github.com/little-quokka/python-panasonic-comfort-cloud/blob/master/pcomfortcloud/session.py

**Comment by novamax89** [comment]
Mine doesn’t work anymore either. Waiting for a solution

**Comment by Pierrefogg** [comment]
> There is already a working API but in Python. We need to translate it into Javascript. Does anyone know about it?
> 
> https://github.com/little-quokka/python-panasonic-comfort-cloud/blob/master/pcomfortcloud/session.py

I know nothing about code. But maybe chatgpt could translate it…

**Comment by nanos** [comment]
Sorry, just asking for clarification: is v4 (specifically v4.0.0-beta8) expected to be working?

the README indicates it should, but this issue indicates it shouldn’t, so just seeking clarification as 4.0.0-beta8 isn’t working for me. Thanks.

**Comment by Pierrefogg** [comment]
> Sorry, just asking for clarification: is v4 (specifically v4.0.0-beta8) expected to be working?
> 
> the README indicates it should, but this issue indicates it shouldn’t, so just seeking clarification as 4.0.0-beta8 isn’t working for me. Thanks.

I don't think it is working at the moment (doesn't for me at least).

**Comment by mkz212** [comment]
@nanos @Pierrefogg  No. It doesn't work yet. Panasonic change API and a large part of the plugin needs to be rebuilt and rewritten.

**Comment by nanos** [comment]
Thanks @mkz212 and thanks for taking the time to respond and for your hard work! Really appreciate it!

**Comment by jandersonhill** [comment]
If anyone at Panasonic is monitoring these discussions, it would be a huge step forward for your customers if you could publish a formal API for Comfort Cloud, or offer native integration with Apple HomeKit.

While server bandwidth might be a concern, it wouldn't be difficult to force a hard limit on the number of API calls per indoor unit. Alternatively, charging a small subscription to allow people to use the API could cover the costs.

Not being able to use Apple HomeKit with your new air conditioning units in 2024 seems nuts!

**Comment by mkz212** [comment]
> > There is already a working API but in Python. We need to translate it into Javascript. Does anyone know about it?
> > https://github.com/little-quokka/python-panasonic-comfort-cloud/blob/master/pcomfortcloud/session.py
> 
> I know nothing about code. But maybe chatgpt could translate it…

Thanks for the suggestion but.. We have already tried with different converters but with medium result.

**Comment by mkz212** [comment]
I updated the first entry in this Issue with the "how you can help" section if anyone was curious and wanted to help.

**Comment by mkz212** [comment]
beta 15 released! Enyone can help? 

This is manual how to check new API: https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/blob/master/docs/intercept-traffic.md

This is working example, but in Python (this plugin is in Javascript): https://github.com/craibo/panasonic_cc/blob/master/custom_components/panasonic_cc/pcomfortcloud/panasonicsession.py

**Comment by mkz212** [comment]
@heldchen Can we continue here? Can you take a quick look? https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/blob/master/src/comfort-cloud.ts

Log:

<img width="1189" alt="Zrzut ekranu 2024-06-20 o 15 45 55" src="https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/assets/82271669/dc37f73f-aded-4587-a04a-b90620b173e4">

**Comment by heldchen** [comment]
the `Location:` header returned from the server only contains the url path - as you're not in a real browser, you need to prepend the host to get the full url in your javascript code.

**Comment by mkz212** [comment]
> the `Location:` header returned from the server only contains the url path - as you're not in a real browser, you need to prepend the host to get the full url in your javascript code.

You probably mean step 2? But that doesn't even reach to this step. See line 116.

**Comment by heldchen** [comment]
> your code tries to redirect to `/login?...` instead of `https://authglb.digital.panasonic.com/login?...` which of course will fail.

**Comment by mkz212** [comment]
> > your code tries to redirect to `/login?...` instead of `https://authglb.digital.panasonic.com/login?...` which of course will fail.

Can you make Pull Request please? Or write me here what to fix? I look at the code from Python and I don't see the difference.

**Comment by heldchen** [comment]
I am not a javascript programmer, no.

**Comment by mkz212** [comment]
> I am not a javascript programmer, no. I guess you'll have to wait for someone with experience to complete this.

Ok thank you!

**Comment by nanos** [comment]
@mkz212 Any advice on how to set up a dev environment to test this?

**Comment by mkz212** [comment]
> @mkz212 Any advice on how to set up a dev environment to test this?

Easiest way is to install current beta and edit files via FTP. 

### For Raspberry Pi: 
You can activate FTP by going to Homebridge Terminal and type 'sudo hb-config'. There go to Configure OS / System options / SSH and activate it (yes, SSH will also activate FTP). You can also change password go to Configure OS / System options / Password.

### FTP:
host: IP of your raspberry
username: pi
port: 22 (using sftp)

root folder of the plugin: /var/lib/homebridge/node_modules/homebridge-panasonic-ac-platform

**Comment by mkz212** [comment]
Error for fix now is that axios response url is relative instead of absolute (full path).

**Comment by nanos** [comment]
I believe that the root cause of the error is that axios follows the redirect automatically (despite specifying `maxRedirects: 0`) and in so doing it redirects to the relative URL rather than the absolute, and that then triggers the 'catch' clause which rejects the Promise.

Whilst on the face of it this makes no sense, this is the only logical explanation I have for the function triggering both the `then` and `catch` methods.

I do know that this is how axios works in the browser (because xhr requests are controlled by the browser, and the browser therefore follows the redirect), but I'm surprised it's doing that in node as well.

Happy to be proven wron though.

**Comment by mkz212** [comment]
> I believe that the root cause of the error is that axios follows the redirect automatically (despite specifying `maxRedirects: 0`) and in so doing it redirects to the relative URL rather than the absolute, and that then triggers the 'catch' clause which rejects the Promise.
> 
> Whilst on the face of it this makes no sense, this is the only logical explanation I have for the function triggering both the `then` and `catch` methods.
> 
> I do know that this is how axios works in the browser (because xhr requests are controlled by the browser, and the browser therefore follows the redirect), but I'm surprised it's doing that in node as well.
> 
> Happy to be proven wron though.

Yes, when i remove catch block it will go to the step 2 and there i have error. I will release next beta and you can see.

**Comment by nanos** [comment]
So what does kinda work is to provide a `baseUrl` parameter:

```js
await axios.request({
      method: 'get',
      url: 'https://authglb.digital.panasonic.com/authorize',
      headers: {
        'user-agent': 'okhttp/4.10.0',
      },
      params: {
        'scope': 'openid offline_access comfortcloud.control a2w.control',
        'audience': 'https://digital.panasonic.com/' + APP_CLIENT_ID +'/api/v1/',
        'protocol': 'oauth2',
        'response_type': 'code',
        'code_challenge': code_challenge,
        'code_challenge_method': 'S256',
        'auth0Client': AUTH_0_CLIENT,
        'client_id': APP_CLIENT_ID,
        'redirect_uri': REDIRECT_URI,
        'state': state,
      },
      maxRedirects: 0,
      validateStatus: status => (status >= 200 && status < 300) || status === 302,
      baseUrl: 'https://authglb.digital.panasonic.com',
    })
```

(note that last line)

That will then mean that the redirect is followed correctly. Axios itself does then follow the redirect (So step 2 likely wouldn't be needed), but I'm getting an error then:


```
[20/06/2024, 16:44:02] [Homebridge Panasonic AC Platform] -------------------------- STEP 1 --------------------------
[20/06/2024, 16:44:02] [Homebridge Panasonic AC Platform] Comfort Cloud - authorize
[20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Comfort Cloud - authorize - Success
[20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Found. Redirecting to https://authst-pn.digital.panasonic.com/error/?client_id={redacted}&connection=&lang=&error=invalid_request&error_description=Missing%20required%20parameter%3A%20client_id&tracking={redacted}
[20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Comfort Cloud - authorize - Error
[20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] {
[20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Error: Invalid URL
```

**Comment by mkz212** [comment]
> So what does kinda work is to provide a `baseUrl` parameter:
> 
> ```js
> await axios.request({
>       method: 'get',
>       url: 'https://authglb.digital.panasonic.com/authorize',
>       headers: {
>         'user-agent': 'okhttp/4.10.0',
>       },
>       params: {
>         'scope': 'openid offline_access comfortcloud.control a2w.control',
>         'audience': 'https://digital.panasonic.com/' + APP_CLIENT_ID +'/api/v1/',
>         'protocol': 'oauth2',
>         'response_type': 'code',
>         'code_challenge': code_challenge,
>         'code_challenge_method': 'S256',
>         'auth0Client': AUTH_0_CLIENT,
>         'client_id': APP_CLIENT_ID,
>         'redirect_uri': REDIRECT_URI,
>         'state': state,
>       },
>       maxRedirects: 0,
>       validateStatus: status => (status >= 200 && status < 300) || status === 302,
>       baseUrl: 'https://authglb.digital.panasonic.com',
>     })
> ```
> 
> (note that last line)
> 
> That will then mean that the redirect is followed correctly. Axios itself does then follow the redirect (So step 2 likely wouldn't be needed), but I'm getting an error then:
> 
> ```
> [20/06/2024, 16:44:02] [Homebridge Panasonic AC Platform] -------------------------- STEP 1 --------------------------
> [20/06/2024, 16:44:02] [Homebridge Panasonic AC Platform] Comfort Cloud - authorize
> [20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Comfort Cloud - authorize - Success
> [20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Found. Redirecting to https://authst-pn.digital.panasonic.com/error/?client_id={redacted}&connection=&lang=&error=invalid_request&error_description=Missing%20required%20parameter%3A%20client_id&tracking={redacted}
> [20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Comfort Cloud - authorize - Error
> [20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] {
> [20/06/2024, 16:44:03] [Homebridge Panasonic AC Platform] Error: Invalid URL
> ```

on the axios page in docs there is baseURL not baseUrl . Thank you! I will test!

**Comment by nanos** [comment]
Quite right. My apologies.

**Comment by mkz212** [comment]
@nanos 

You can enable debug (in plugin config and Homebridge settings) - then you will have more info in logs.

So far i see that w commenting or removing Promises in catch block will pass to next steps. 

But in step 2 i have status 302 (should be 200).

For me baseURL and baseUrl doesn't work..

**Comment by nanos** [comment]
Yeah, there is still something weird going on here. I don't think step 2 is expected to work in this instance, because, having followed the redirect, there is no 302, and thus no `Location` header, and thus this won't work.

**Comment by nanos** [comment]
but you are correct that i originally misread the logs: axios still appears to follow a relative redirect to `/login?state...` which is not a valid URL.

I don't know how to fix this though, and I need to make dinner for my kids now, so I'll be away from my computer for a few hours.

**Comment by mkz212** [comment]
> Yeah, there is still something weird going on here. I don't think step 2 is expected to work in this instance, because, having followed the redirect, there is no 302, and thus no `Location` header, and thus this won't work.

'location' - small letter  i fixed this is newest beta

**Comment by mkz212** [comment]
> but you are correct that i originally misread the logs: axios still appears to follow a relative redirect to `/login?state...` which is not a valid URL.
> 
> I don't know how to fix this though, and I need to make dinner for my kids now, so I'll be away from my computer for a few hours.

I'm also finishing for today, I've been working on it since the morning... I am grateful to you for your help and if you work out something more, write here or do Pull Request. In general, the working code is in a Python version, but there they do not use axios only something from Python, there may also be problems.

**Comment by nanos** [comment]
OK, so I made some progress - or maybe not?

I've change the `axios.request` call in line 90 as follows:

1. Removed `maxRedirects: 0,`
2. Added a `beforeRedirect` option that simply prepends the host name to the path (and extracts the csrf cookie):

The whole thing then looks as follows:

```js
await axios.request({
      method: 'get',
      url: 'https://authglb.digital.panasonic.com/authorize',
      headers: {
        'user-agent': 'okhttp/4.10.0',
      },
      params: {
        'scope': 'openid offline_access comfortcloud.control a2w.control',
        'audience': 'https://digital.panasonic.com/' + APP_CLIENT_ID +'/api/v1/',
        'protocol': 'oauth2',
        'response_type': 'code',
        'code_challenge': code_challenge,
        'code_challenge_method': 'S256',
        'auth0Client': AUTH_0_CLIENT,
        'client_id': APP_CLIENT_ID,
        'redirect_uri': REDIRECT_URI,
        'state': state,
      },
      beforeRedirect: (options, { headers }) => {
        if (headers.location && headers.location.startsWith('/')) {
          headers.location = 'https://authglb.digital.panasonic.com' + headers.location;
          this.log.debug(JSON.stringify(headers));
          this.csrf = (headers['set-cookie'] as string[])
            .find(cookie => cookie.includes('_csrf'))
            ?.match(new RegExp('^_csrf=(.+?);'))
            ?.[1];
        }
      },
      validateStatus: status => (status >= 200 && status < 300) || status === 302,
    })
```

This still produces an error, presumably because it's now missing something else. But if we get that to work, then maybe we can skip step 2 altogether? I genuinely don't know whether this is helpful, or a complete wild goose chase though!

**Comment by mkz212** [comment]
@nanos 

I think it can be as heldchen wrote : "the python code uses a browsing session class, i.e. it automatically keeps track of cookies & some headers (like referrer) between requests. if your javascript helper does not do that, you'll need to always ensure that the correct cookies, headers etc are set exactly as the service expects. you can open the /authorize url in a browser and check the dev-tools to see what is in each of the requests & responses."

also check: https://github.com/craibo/panasonic_cc/blob/master/custom_components/panasonic_cc/pcomfortcloud/panasonicsession.py

You can check there how the flow should look and what status response each request should have.

**Comment by nanos** [comment]
I think the canonical answer to cookies/session with axios is https://github.com/3846masa/axios-cookiejar-support.

**Comment by mkz212** [comment]
@nanos

- i tried axios instance: https://xcelore.com/mastering-http-requests-with-axios-instances-a-comprehensive-guide/
- I tried axios beforeRedirect
- I tried axios withCredentials

Everytime with no luck. Maybe need to use something other than axios?

**Comment by nanos** [comment]
Ok, I think I just got a breakthrough! Now getting all the way through to step 5 🎉 (although I’m still getting an error there)

The problem was indeed missing cookies. 

need to do some tidying up, then will file a PR.

**Comment by mkz212** [comment]
> Ok, I think I just got a breakthrough! Now getting all the way through to step 5 🎉 (although I’m still getting an error there)
> 
> The problem was indeed missing cookies.
> 
> need to do some tidying up, then will file a PR.

Wow! That's just good news in the morning! 🚀I'm very curious what was wrong there

**Comment by nanos** [comment]
Yes, I managed to include that in PR #360.

Sadly, at step 6 I'm being blocked by CloudFront - not sure if there is anything we can do about this from our end? Maybe different UA, maybe just wait half an hour?

**Comment by nanos** [comment]
Just FYI: The way cookie jar works is as follows:

1) import it:

```js
import { wrapper } from 'axios-cookiejar-support';
import { CookieJar } from 'tough-cookie';
```

2) Create a cookie jar:
```js
const jar = new CookieJar();
const client = wrapper(axios.create({ jar }));
```

3) wherever you would've previously used `axios` just use `client` (drop in replacement):
```js
client.request('https://...')
```

**Comment by nanos** [comment]
OK, the reason for the CloudFlare block is that we forgot a step in between, and passed on wrong UA headers. Just filed PR #361 which allows me to complete the login flow all the way.

**Comment by nanos** [comment]
Now ending up with this error:

```
[21/06/2024, 09:25:47] [Homebridge Panasonic AC Platform] -------------------------- STEP 9 --------------------------
[21/06/2024, 09:25:47] [Homebridge Panasonic AC Platform] Successfully logged in to Comfort Cloud.
[21/06/2024, 09:25:47] [Homebridge Panasonic AC Platform] Discovering devices on Comfort Cloud.
[21/06/2024, 09:25:47] [Homebridge Panasonic AC Platform] Comfort Cloud: getDevices()
[21/06/2024, 09:25:48] [Homebridge Panasonic AC Platform] Comfort Cloud - getDevices(): Error
[21/06/2024, 09:25:49] [Homebridge Panasonic AC Platform] {
[21/06/2024, 09:25:49] [Homebridge Panasonic AC Platform] An error occurred during device discovery. Turn on debug mode for more information.
[21/06/2024, 09:25:49] [Homebridge Panasonic AC Platform] Comfort Cloud - getDevices(): Error
```

As this is part of the code that I have not previously explored I will bow out here for now. Need to get some work done...

**Comment by mkz212** [comment]
Wow! Thank you! 🚀

I will try to solve this. I think they also had a problem with it in the Python plugin.

**Comment by mkz212** [comment]
@nanos I did some fixes and cleanings and released beta.20. Maybe in getDevice() and refreshToken() you also need to access cookie jar in request?

**Comment by nanos** [comment]
Cheers. Just filed another PR.

Remember that the header for the access token always needs to be `X-User-Authorization-V2: Bearer {token}`, so needs to have the word `Bearer` and a space prefixed to the actual token

**Comment by nanos** [comment]
IT WORKS!!!! 
<img width="475" alt="image" src="https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/assets/5182595/ac949da2-55ef-4b48-b007-59b53b29ca66">

**Comment by mkz212** [comment]
@nanos Many many THANKS! We couldn't do it without you! 🚀👍

1000 likes and comment from Ryan Reynold and i will release public! 😜

And seriously, let's check today and tomorrow if everything works, e.g. the token is refreshed every 24 hours.

**Comment by nanos** [comment]
Yes, I think the token refresh is the big wildcard, isn't it. But we certainly made big progress today!

**Comment by nanos** [comment]
lol, yeah, no. Refresh token won't work. Let me fix that real quick

**Comment by mkz212** [comment]
> lol, yeah, no. Refresh token won't work. Let me fix that real quick

Once again many many thanks!!! 👍

**Comment by mkz212** [comment]
> the `Location:` header returned from the server only contains the url path - as you're not in a real browser, you need to prepend the host to get the full url in your javascript code. your code tries to redirect to `/login?...` instead of `https://authglb.digital.panasonic.com/login?...` which of course will fail.
> 
> 1. construct & fetch `/authorize` url (`GET`)
> 2. parse the response (redirect target, cookies)
> 3. construct & fetch `/login` url (`GET`)
> 4. parse the response (`state` value, `csrf` value, cookies)
> 5. construct & fetch `/userpassword/login` url (`POST`)
> 6. parse the response (form `action` and its hidden input fields)
> 7. construct & fetch the callback url based on the previous' step form values (`POST`)
> 8. parse the response (redirect target)
> 9. construct & fetch the target url (`GET`)
> 10. parse the response (`code` parameter you need to complete the oauth challenge)
> 11. construct & fetch `/oauth/token` url (`POST`)
> 12. now you have the token, yay!
> 
> the exact headers, cookies, url parameters & post data you can find in the code block https://github.com/lostfields/python-panasonic-comfort-cloud/blob/87afbd27b7d0f205d3dbbfde0e437b1a9937ec08/pcomfortcloud/panasonicsession.py#L88 (line 88 to 252).
> 
> the python code uses a browsing session class, i.e. it automatically keeps track of cookies & some headers (like referrer) between requests. if your javascript helper does not do that, you'll need to always ensure that the correct cookies, headers etc are set _exactly_ as the service expects. you can open the /authorize url in a browser and check the dev-tools to see what is in each of the requests & responses.

Thank you! It helped a lot and the plugin is already working! 👍

**Comment by nanos** [comment]
Just on another note in case you are not aware: 

The main thing that really held me back yesterday was that I needed to do development on the home bridge server, where I had only logs and couldn't attach a debugger.

I have created a `src/test.ts` file with the following content:

```js
import { Logger } from 'homebridge';
import ComfortCloudApi from './comfort-cloud';
import { PanasonicPlatformConfig } from './types';
import PanasonicPlatformLogger from './logger';
import PanasonicPlatform from './platform';
import { HomebridgeAPI } from 'homebridge/lib/api';

const config = {
  'email': '{email}',
  'password': '{password}',
  'debugMode': true,
};
const logger = new PanasonicPlatformLogger(
  console as unknown as Logger,
  2,
);
const api = new ComfortCloudApi(
  config as PanasonicPlatformConfig,
  logger,
);

const platform = new PanasonicPlatform(
  console as unknown as Logger,
  config as PanasonicPlatformConfig,
  new HomebridgeAPI,
);

platform.loginAndDiscoverDevices();

// test refresh token
api.login().then(() => {
  api.refreshToken();
});
```

This can then be run using `npx tsx ./src/test.ts` (no need to `build` before) from the command line from your local machine, where you can attach a debugger, and overall just develop much quicker and without needing releases, build steps etc.

Obviously I haven't committed this file, as it contains my credentials, but if you are interested I can file a PR with instructions on how to set this up in your contribution guide?

**Comment by mkz212** [comment]
@nanos From now on, the API works so that in the Panasonic Comfort Cloud application (iOS / Android) there is a hardcoded client id. It is different for each version of the application. Here is the manual: https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/blob/master/docs/intercept-traffic.md . Do you know about it?

**Comment by mkz212** [comment]
> Just on another note in case you are not aware:
> 
> The main thing that really held me back yesterday was that I needed to do development on the home bridge server, where I had only logs and couldn't attach a debugger.
> 
> I have created a `src/test.ts` file with the following content:
> 
> ```js
> import { Logger } from 'homebridge';
> import ComfortCloudApi from './comfort-cloud';
> import { PanasonicPlatformConfig } from './types';
> import PanasonicPlatformLogger from './logger';
> import PanasonicPlatform from './platform';
> import { HomebridgeAPI } from 'homebridge/lib/api';
> 
> const config = {
>   'email': '{email}',
>   'password': '{password}',
>   'debugMode': true,
> };
> const logger = new PanasonicPlatformLogger(
>   console as unknown as Logger,
>   2,
> );
> const api = new ComfortCloudApi(
>   config as PanasonicPlatformConfig,
>   logger,
> );
> 
> const platform = new PanasonicPlatform(
>   console as unknown as Logger,
>   config as PanasonicPlatformConfig,
>   new HomebridgeAPI,
> );
> 
> platform.loginAndDiscoverDevices();
> 
> // test refresh token
> api.login().then(() => {
>   api.refreshToken();
> });
> ```
> 
> This can then be run using `npx tsx ./src/test.ts` (no need to `build` before) from the command line from your local machine, where you can attach a debugger, and overall just develop much quicker and without needing releases, build steps etc.
> 
> Obviously I haven't committed this file, as it contains my credentials, but if you are interested I can file a PR with instructions on how to set this up in your contribution guide?

I don't know much about it. I am not a professional programmer. I never needed such things. But of course, any help is always welcome!

**Comment by nanos** [comment]
> From now on, the API works so that in the Panasonic Comfort Cloud application (iOS / Android) there is a hardcoded client id. It is different for each version of the application.

Wow, what a pain! This might make this really really hard to keep on top of updates, as you presumably need to decompile and extract these on every release.

**Comment by mkz212** [comment]
> > From now on, the API works so that in the Panasonic Comfort Cloud application (iOS / Android) there is a hardcoded client id. It is different for each version of the application.
> 
> Wow, what a pain! This might make this really really hard to keep on top of updates, as you presumably need to decompile and extract these on every release.

That's right. Panasonic does not have an official API for plugins, only for its application. The average user runs applications up to several times a day. And such a plug-in even asks you for a few seconds. Ours asks every 10 minutes when the device is turned on and 60 minutes when it is turned off. But some plugins generate a lot of traffic on the servers. That's probably why he's fighting it so much.

**Comment by mkz212** [comment]
> > From now on, the API works so that in the Panasonic Comfort Cloud application (iOS / Android) there is a hardcoded client id. It is different for each version of the application.
> 
> Wow, what a pain! This might make this really really hard to keep on top of updates, as you presumably need to decompile and extract these on every release.

I updated the manual. Now it's easy to read APP_CLIENT_ID, but I don't know where to find AUTH_0_CLIENT.

https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/blob/master/docs/app.md

**Comment by mkz212** [comment]
> > From now on, the API works so that in the Panasonic Comfort Cloud application (iOS / Android) there is a hardcoded client id. It is different for each version of the application.
> 
> Wow, what a pain! This might make this really really hard to keep on top of updates, as you presumably need to decompile and extract these on every release.

I updated the manual. Now it's easy to read APP_CLIENT_ID and AUTH_0_CLIENT.

https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/blob/master/docs/app.md

**Comment by marc2016** [comment]
Thank you very much for your work. This helped me update my package! :-)

**Comment by mkz212** [comment]
> Thank you very much for your work. This helped me update my package! :-)

I'm glad it works! 👍 Let's just hope they don't change something so much again. 🙄

**Comment by jandersonhill** [comment]
Good thinking!

From: mkz212 ***@***.***>
Date: Thursday, 20 June 2024 at 06:32
To: homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform ***@***.***>
Cc: Jeremy ***@***.***>, Comment ***@***.***>
Subject: Re: [homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform] Plugin doesn't work. Panasonic changed API. (Issue #330)

I updated the first entry in this Issue with the "how you can help" section if anyone was curious and wanted to help.
